_posts/2021-02-04-justctf21-rusty.md

```diff
@@ -0,0 +1,219 @@
+the project as a separate file.
+
+#### Analyzing the DOS stub
+
+Since I cut out the header region without any PE/MZ header, some non-conventional steps are needed when
+importing it into Ghidra.
+
+![Import raw binary 1](/assets/images/posts/justctf21_rusty/rusty-import-dos-stub.bin-1.png "Import raw binary in Ghidra #1")
+
+![Import raw binary 2](/assets/images/posts/justctf21_rusty/rusty-import-dos-stub.bin-2.png "Import raw binary in Ghidra #2")
+
+To find out where the first instruction is, we need to go back and inspect the MZ header.
+
+`e_ip` (Initial IP value) is set to `0x8a` - but what does it offset of?
+
+Two DOS terms we need to know about:
+
+* A _paragraph_ is `16` bytes (`0x10`)
+* A _page_ is `512` bytes (`0x200`)
+
+Turns out that we need to take page alignment into account. Since the header up to the `e_program`
+section is `0x40` bytes, the current page will be padded with `0x200 - 0x40 = 0x1c0 (448)` bytes
+before the page(s) for the working set of the executable.
+
+![Ghidra byte listing](/assets/images/posts/justctf21_rusty/rusty-dos-stub-bin-listing-1.png "Ghidra raw listing of rusty DOS stub")
+
+At `0x1c0`, we discover the bytes `This program cannot be run in DOS Mode.$`. Notice the
+`$` at the end? Strings handled by the syscalls of `INT 0x21` terminated by a dollar - not by `0`
+as you might be used to - or Pascal-style, with the string length stated as a prefix byte.
+
+At `0x1c0 + 0x8 = 0x24a`, we find `0xfc` which translates to the `CLD` (Clear Direction Flag) in 8086 assembly.
+
+![Ghidra assembly listing](/assets/images/posts/justctf21_rusty/rusty-dos-stub-bin-listing-2.png "Ghidra assembly listing of rusty DOS stub")
+
+Since demo scene-related coding was something I did during the mid-90'es, I quickly realize that I'm
+familiar with most of the stuff going on.
+
+Before analyzing the binary further, I installed `dosbox-debug` and fired it up.
+
+I mounted a local directory containing `rusty.exe`:
+
+```shell
+Z:\>MOUNT r /home/larsw/justctf/re/rusty
+Z:\>r:
+R:\>RUSTY.EXE
+```
+
+![Flames effect](/assets/images/posts/justctf21_rusty/rusty-flames.png "Flames effect!")
+
+Yay! Instant déjà vu!
+
+Before exiting the program (with either ESC or Enter), I noticed that for each key I pressed,
+a asterisk was displayed and merged into the flame effect.
+
+Back to Ghidra. As I mentioned earlier, most of the decompiled assembly made sense to me, so
+I started to clean up/rename functions until I felt that I had identified all the key components.
+Apart from setting up "mode 13h" (320x200/8-bit) and a nice palette (doing I/O on port 0x3c7/0x3c8),
+there was also the standard `wait_for_vsync` routine, and of course the logic for generating some
+pseudo random bytes and creating three copies of them on the bottom of the screen, before doing passes
+to smooth out the rows above. All this was done in an never-ending loop with a check for keyboard status
```

Comment:
All this was done in an never-ending loop

Skal vel være "a never ending loop", ikke "an"